File: catalogue-application/src/main/java/io/quartic/catalogue/inmemory/InMemoryStorageBackend.java
package io.quartic.catalogue.inmemory;

import com.codahale.metrics.health.HealthCheck;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Maps;
import io.quartic.catalogue.StorageBackend;
import io.quartic.catalogue.api.model.DatasetConfig;
import io.quartic.catalogue.api.model.DatasetCoordinates;

import java.io.IOException;
import java.util.Map;

public class InMemoryStorageBackend implements StorageBackend {
    private final Map<DatasetCoordinates, DatasetConfig> datasets = Maps.newConcurrentMap();

    // TODO
    @Override
    public DatasetConfig get(DatasetCoordinates coords) throws IOException {
        return datasets.get(coords);
    }

    // TODO
    @Override
    public void put(DatasetCoordinates coords, DatasetConfig config) throws IOException {
        datasets.put(coords, config);
    }

    // TODO
    @Override
    public void remove(DatasetCoordinates coords) throws IOException {
        datasets.remove(coords);
    }

    // TODO
    @Override
    public boolean contains(DatasetCoordinates coords) throws IOException {
        return datasets.containsKey(coords);
    }

    // TODO
    @Override
    public Map<DatasetCoordinates, DatasetConfig> getAll() throws IOException {
        return ImmutableMap.copyOf(datasets);
    }

    @Override
    public HealthCheck.Result healthCheck() {
        return HealthCheck.Result.healthy(); // can't really be any other way
    }
}

File: catalogue/src/main/kotlin/io/quartic/catalogue/inmemory/InMemoryStorageBackend.kt
package io.quartic.catalogue.inmemory

import com.codahale.metrics.health.HealthCheck.Result
import io.quartic.catalogue.StorageBackend
import io.quartic.catalogue.api.model.DatasetConfig
import io.quartic.catalogue.api.model.DatasetCoordinates

class InMemoryStorageBackend : StorageBackend {
    private val datasets = mutableMapOf<DatasetCoordinates, DatasetConfig>()

    @Synchronized
    override fun get(coords: DatasetCoordinates): DatasetConfig? = datasets[coords]

    @Synchronized
    override fun set(coords: DatasetCoordinates, config: DatasetConfig) {
        datasets[coords] = config
    }

    @Synchronized
    override fun contains(coords: DatasetCoordinates) = coords in datasets

    @Synchronized
    override fun remove(coords: DatasetCoordinates) {
        datasets.remove(coords)
    }

    @Synchronized
    override fun getAll(): Map<DatasetCoordinates, DatasetConfig> = HashMap(datasets)

    @Synchronized
    override fun healthCheck(): Result = Result.healthy() // can't really be any other way
}
